svm/src/account_loader.rs

```diff
@@ -2227,4 +2371,271 @@ mod tests {
+                ixns,
+                account1_size
+                    + account2_size
+                    + program1_size
+                    + program2_size
+                    + bpf_loader_size
+                    + upgradeable_loader_size
+                    + fee_payer_size,
+            );
+
+            // ordinary owners not counted
+            let ixns = vec![Instruction::new_with_bytes(
+                program1,
+                &[],
+                vec![account_meta(account2, false)],
+            )];
+            test_data_size(
+                ixns,
+                account2_size + program1_size + bpf_loader_size + fee_payer_size,
+            );
+
+            // program and loader counted once
+            let ixns = vec![
+                Instruction::new_with_bytes(program1, &[], vec![]),
+                Instruction::new_with_bytes(program1, &[], vec![]),
+            ];
+            test_data_size(ixns, program1_size + bpf_loader_size + fee_payer_size);
+
+            // native loader not counted if loader
+            let ixns = vec![Instruction::new_with_bytes(bpf_loader::id(), &[], vec![])];
+            test_data_size(ixns, bpf_loader_size + fee_payer_size);
+
+            // native loader counted if instruction
+            let ixns = vec![Instruction::new_with_bytes(
+                bpf_loader::id(),
+                &[],
+                vec![account_meta(native_loader::id(), false)],
+            )];
+            test_data_size(ixns, bpf_loader_size + native_loader_size + fee_payer_size);
+
+            // loader counted twice if included in instruction
+            let ixns = vec![Instruction::new_with_bytes(
+                program1,
+                &[],
+                vec![account_meta(bpf_loader::id(), false)],
+            )];
+            test_data_size(ixns, program1_size + bpf_loader_size * 2 + fee_payer_size);
+
+            // cover that case with multiple loaders to be sure
+            let ixns = vec![
+                Instruction::new_with_bytes(
+                    program1,
+                    &[],
+                    vec![
+                        account_meta(bpf_loader::id(), false),
+                        account_meta(bpf_loader_upgradeable::id(), false),
+                    ],
+                ),
+                Instruction::new_with_bytes(program2, &[], vec![account_meta(account1, false)]),
+                Instruction::new_with_bytes(
+                    bpf_loader_upgradeable::id(),
+                    &[],
+                    vec![account_meta(account1, false)],
+                ),
+            ];
+            test_data_size(
+                ixns,
+                account1_size
+                    + program1_size
+                    + program2_size
+                    + bpf_loader_size * 2
+                    + upgradeable_loader_size * 2
+                    + fee_payer_size,
+            );
+
+            // loader counted twice even if included first
```

Comment:
do we have concerns on the order of instructions affecting load size?
if so, wdyt if adding a loop over instruction permutations in `test_data_size`?

Reply:
the only thing i wanted to test with order is loaders, because loader sizes are added in a separate pass, when program indicies are gathered. it does something like "have we appended this loader after the main accounts, if not count and append it." the main loop in the loader gets account keys for the whole message, so order doesnt matter